python/lsst/afw/detection/peak.cc

```diff
@@ -35,40 +37,75 @@ namespace lsst {
+using PyPeakTable = py::class_<PeakTable, std::shared_ptr<PeakTable>, table::BaseTable>;
+using PyPeakCatalog = py::class_<table::CatalogT<PeakRecord>,
+                                 std::shared_ptr<table::CatalogT<PeakRecord>>>;
+
+/**
+Declare constructors and member and static functions for a pybind11 PeakRecord
+*/
+void declarePeakRecord(PyPeakRecord & cls) {
+    table::pybind11::addCastFrom<table::BaseRecord, PeakRecord>(cls);
+
+    cls.def("getTable", &PeakRecord::getTable);
+    cls.def("getId", &PeakRecord::getId);
+    cls.def("setId", &PeakRecord::setId);
+    cls.def("getIx", &PeakRecord::getIx);
+    cls.def("getIy", &PeakRecord::getIy);
+    cls.def("setIx", &PeakRecord::setIx);
+    cls.def("setIy", &PeakRecord::setIy);
+    cls.def("getI", &PeakRecord::getI);
+    cls.def("getCentroid", (afw::geom::Point2I (PeakRecord::*)(bool) const) &PeakRecord::getCentroid);
+    cls.def("getCentroid", (afw::geom::Point2D (PeakRecord::*)() const) &PeakRecord::getCentroid);
+    cls.def("getFx", &PeakRecord::getFx);
+    cls.def("getFy", &PeakRecord::getFy);
+    cls.def("setFx", &PeakRecord::setFx);
+    cls.def("setFy", &PeakRecord::setFy);
+    cls.def("getF", &PeakRecord::getF);
+    cls.def("getPeakValue", &PeakRecord::getPeakValue);
+    cls.def("setPeakValue", &PeakRecord::setPeakValue);
+}
 
-    /* Member types and enums */
+/**
+Declare constructors and member and static functions for a pybind11 PeakTable
+*/
+void declarePeakTable(PyPeakTable & cls) {
+    table::pybind11::addCastFrom<table::BaseTable, PeakTable>(cls);
 
-    /* Constructors */
+    cls.def_static("make", &PeakTable::make, "schema"_a, "forceNew"_a=false);
+    cls.def_static("makeMinimalSchema", &PeakTable::makeMinimalSchema);
+    cls.def_static("checkSchema", &PeakTable::checkSchema, "schema"_a);
+    cls.def("getIdFactory",
+            (std::shared_ptr<table::IdFactory> (PeakTable::*)()) &PeakTable::getIdFactory);
+    cls.def("setIdFactory", &PeakTable::setIdFactory, "factory"_a);
+    cls.def_static("getIdKey", &PeakTable::getIdKey);
+    cls.def_static("getIxKey", &PeakTable::getIxKey);
+    cls.def_static("getIyKey", &PeakTable::getIyKey);
+    cls.def_static("getFxKey", &PeakTable::getFxKey);
+    cls.def_static("getFyKey", &PeakTable::getFyKey);
+    cls.def_static("getPeakValueKey", &PeakTable::getPeakValueKey);
+}
+
+}  // lsst::afw::detection::<anonymous>
 
-    /* Operators */
+PYBIND11_PLUGIN(_peak) {
+    py::module mod("_peak", "Python wrapper for afw _peak library");
+
+    /* Module level */
+    PyPeakRecord clsPeakRecord(mod, "PeakRecord");
+    PyPeakTable clsPeakTable(mod, "PeakTable");
+    PyPeakCatalog clsPeakCatalog(mod, "PeakCatalog", py::dynamic_attr());
 
     /* Members */
+    declarePeakRecord(clsPeakRecord);
+    declarePeakTable(clsPeakTable);
+    table::pybind11::declareCatalog(clsPeakCatalog);
+
+    clsPeakCatalog.attr("Record") = clsPeakRecord;
+    clsPeakCatalog.attr("Table") = clsPeakTable;
```

Comment:
Thanks. Personally I hate the fact that it is so easy to forget, and if forgotten, then casting will do the wrong thing. But I could not find a safe way to make declareCatalog handle it: I could pass the python record and table classes as argument, but I could not think of any way of detecting if they were passed in the wrong order!